consul/server_manager.go

```diff
@@ -0,0 +1,212 @@
+package consul
+
+import (
+	"math/rand"
+	"sync/atomic"
+	"time"
+
+	"github.com/hashicorp/consul/lib"
+)
+
+type consulServerEventTypes int
+
+const (
+	// consulServersNodeJoin is used to notify of a new consulServer.
+	// The primary effect of this is a reshuffling of consulServers and
+	// finding a new preferredServer.
+	consulServersNodeJoin = iota
+
+	// consulServersRebalance is used to signal we should rebalance our
+	// connection load across servers
+	consulServersRebalance
+
+	// consulServersRPCError is used to signal when a server has either
+	// timed out or returned an error and we would like to have the
+	// server manager find a new preferredServer.
+	consulServersRPCError
+)
+
+// serverCfg is the thread-safe configuration structure that is used to
+// maintain the list of consul servers in Client.
+//
+// NOTE(sean@): We are explicitly relying on the fact that this is copied.
+// Please keep this structure light.
+type serverConfig struct {
+	// servers tracks the locally known servers
+	servers []*serverParts
+
+	// Timer used to control rebalancing of servers
+	rebalanceTimer *time.Timer
```

Comment:
My understanding is yes, it's safe to copy a pointer to the timer, but that understanding is derived by what I know of the Go runtime with C++ colored glasses.  I'll see if I can confirm, but think that it's safe because it's word-sized.